src/librustc_traits/lowering.rs

```diff
@@ -260,6 +276,40 @@ fn implied_bound_from_trait<'a, 'tcx>(
     }))
 }
 
+fn wellformed_from_bound<'a, 'tcx>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    trait_pred: ty::TraitPredicate<'tcx>,
+    where_clauses: &[ty::Predicate<'tcx>],
+) -> iter::Once<Clause<'tcx>> {
+    // Rule WellFormed-TraitRef
+    //
+    // For each where clause WC:
+    // forall<Self, P1..Pn> {
+    //   WellFormed(Self: Trait<P1..Pn>) :- Implemented(Self: Trait<P1..Pn>) && WellFormed(WC)
+    // }
+
+    // WellFormed(Self: Trait<P1..Pn>)
+    let wellformed_trait = DomainGoal::WellFormed(WhereClauseAtom::Implemented(trait_pred));
+    // Implemented(Self: Trait<P1..Pn>)
+    let impl_trait = ty::Binder::dummy(DomainGoal::Holds(WhereClauseAtom::Implemented(trait_pred)));
+    // WellFormed(WC)
+    let wellformed_wcs = where_clauses
+        .into_iter()
+        .map(|wc| wc.lower().map_bound(|wc| wc.into_wellformed_goal()));
```

Comment:
Actually you're facing the same problem as @tmandry: `Implemented(Self: Trait)` is included in the where clauses, so this means that you are outputting:
`WellFormed(Self: Trait) :- Implemented(Self: Trait), WellFormed(Self: Trait)`
which is unwanted (see your examples below).

I don't know how to deal with this without https://github.com/rust-lang/rust/pull/50183...